f3.go

```diff
@@ -145,412 +78,197 @@ func New(ctx context.Context, manifest manifest.ManifestProvider, ds datastore.D
-		}
-	}
-	m.psLk.Unlock()
-}
-
-func (m *F3) setCertStore(cs *certstore.Store) {
-	m.csLk.Lock()
-	defer m.csLk.Unlock()
-	m.certStore = cs
-	m.client.certStore = cs
-}
+	m.mu.Lock()
+	runner := m.runner
+	m.mu.Unlock()
 
-func (m *F3) setupPubsub(runner *gpbftRunner) error {
-	pubsubTopicName := m.Manifest().PubSubTopic()
-
-	// explicit type to typecheck the anonymous function defintion
-	// a bit ugly but I don't want gpbftRunner to know about pubsub
-	var validator pubsub.ValidatorEx = func(ctx context.Context, pID peer.ID,
-		msg *pubsub.Message) pubsub.ValidationResult {
-
-		var gmsg gpbft.GMessage
-		err := gmsg.UnmarshalCBOR(bytes.NewReader(msg.Data))
-		if err != nil {
-			return pubsub.ValidationReject
-		}
-		validatedMessage, err := runner.ValidateMessage(&gmsg)
-		if errors.Is(err, gpbft.ErrValidationInvalid) {
-			m.log.Debugf("validation error during validation: %+v", err)
-			return pubsub.ValidationReject
-		}
-		if err != nil {
-			m.log.Warnf("unknown error during validation: %+v", err)
-			return pubsub.ValidationIgnore
-		}
-		msg.ValidatorData = validatedMessage
-		return pubsub.ValidationAccept
-	}
-
-	err := m.pubsub.RegisterTopicValidator(pubsubTopicName, validator)
-	if err != nil {
-		return xerrors.Errorf("registering topic validator: %w", err)
+	if runner == nil {
+		m.log.Error("attempted to broadcast message while F3 wasn't running")
+		return
 	}
 
-	topic, err := m.pubsub.Join(pubsubTopicName)
+	err := runner.BroadcastMessage(msg)
 	if err != nil {
-		return xerrors.Errorf("could not join on pubsub topic: %s: %w", pubsubTopicName, err)
+		m.log.Errorf("failed to broadcast message: %+v", err)
 	}
-	m.client.topic = topic
-	return nil
 }
 
-func (m *F3) teardownPubsub(manifest manifest.Manifest) (err error) {
-	m.psLk.Lock()
-	defer m.psLk.Unlock()
-	if m.client.topic != nil {
-		m.msgSub.Cancel()
-		err = multierr.Combine(
-			m.pubsub.UnregisterTopicValidator(manifest.PubSubTopic()),
-			m.client.topic.Close(),
-		)
-		m.client.topic = nil
-	}
-	return err
-}
-
-func (m *F3) initGpbftRunner(ctx context.Context) error {
-	m.mLk.Lock()
-	defer m.mLk.Unlock()
-
-	cs, err := certstore.OpenStore(ctx, m.ds)
-	if err == nil {
-		m.setCertStore(cs)
-	} else if errors.Is(err, certstore.ErrNotInitialized) {
-		err := m.boostrap(ctx)
-		if err != nil {
-			return xerrors.Errorf("failed to boostrap: %w", err)
-		}
-	} else {
-		return xerrors.Errorf("opening certstore: %w", err)
-	}
-
-	m.runner, err = newRunner(m.client.manifest, m.client)
-	if err != nil {
-		return xerrors.Errorf("creating gpbft host: %w", err)
-	}
-
-	m.psLk.Lock()
-	if err := m.setupPubsub(m.runner); err != nil {
-		return xerrors.Errorf("setting up pubsub: %w", err)
-	}
-	m.msgSub, err = m.client.topic.Subscribe()
-	if err != nil {
-		m.psLk.Unlock()
-		return xerrors.Errorf("subscribing to topic: %w", err)
-	}
-	m.psLk.Unlock()
-	return nil
-}
+func (m *F3) GetLatestCert(ctx context.Context) (*certs.FinalityCertificate, error) {
+	m.mu.Lock()
```

Comment:
I think we can just use `defer` here, it is fast enough and non-blocking in any way.

Reply:
Although we might want to keep it like this for uniformity.

Reply:
Yeah, I mostly wanted to be consistent.